### PR TITLE
Merge jeffbean fork 

### PR DESCRIPTION
**Pull request by jeffbean**
Adding the Reconfig APIs. 

Mostly updating the integration test framework to be more reliable. Some general cleanup and reordering of functions. 

- adding 3.5 reconfig APIs. 
- update testing framework to deal with different versions of ZK (3.4 and 3.5)
- adding makefile for more constant test/build environments
- fixing race in existing test 
- adding testing.T to StartTestCluster to ensure its only used in tests.

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -219,9 +222,11 @@ func Connect(servers []string, sessionTimeout time.Duration, options ...connOpti
 	}
 
 	conn.setTimeouts(int32(sessionTimeout / time.Millisecond))
+	// TODO: This context should be passed in by the caller to be the connection lifecycle context.
+	ctx := context.Background()
```
Comment (nemith) [inline conn.go]: Perhaps this could be context.TODO() instead

**Comment by nemith** [comment]
Can you rebase this?  I think there is some already committed changes in here.

**Comment by jeffbean** [comment]
> Can you rebase this? I think there is some already committed changes in here.

Yes. i can update this weekend. ty for looking

**Comment by nemith** [comment]
This has been rebased and merged manually

**Pull request closed without merging**